docs/guide/misc/grib_memory.rst

```diff
@@ -0,0 +1,91 @@
+
+Reading all the GRIB data into memory
+========================================
+
+We can load :ref:`grib` data fully into memory when we read it as a :ref:`stream <streams>` with the ``read_all=True`` option in :func:`from_source`.
+
+.. code-block:: python
+
+    import earthkit.data
+
+    url = "https://get.ecmwf.int/repository/test-data/earthkit-data/examples/test6.grib"
+    ds = earthkit.data.from_source("url", url, stream=True, read_all=True)
+
+With this the entire ``ds`` fieldlist, including all the fields and the related GRIB handles, are stored in memory.
+
+This technique also works for GRIB data on disk, we just need to read it as a :ref:`stream source <data-sources-stream>`.
+
+.. code-block:: python
+
+    import earthkit.data
+
+    f = open("test6.grib", "rb")
+    ds = earthkit.data.from_source("stream", f, read_all=True)
+
+.. warning::
+
+    Use this option carefully since your data might not fit into memory.
+
+Reading data from disk and partially store it in memory
+===========================================================
+
+When reading :ref:`grib` data from disk as a :ref:`file source <data-sources-file>` it is represented as a fieldlist and loaded lazily. After the (fast) initial scan for field offsets and lengths no actual fields are created and no data is read into memory. When we start using the fieldlist, e.g. by iterating over the fields, accessing data or metadtata etc., the fields will be created on demand and the related GRIB handles will be loaded from disk when needed. Whether this data or part of it stays in memory depends on the following :ref:`settings <settings>`:
+
+- :ref:`store-grib-fields-in-memory <store-grib-fields-in-memory>`
+- :ref:`grib-handle-cache-size <grib-handle-cache-size>`
+- :ref:`use-grib-metadata-cache <use-grib-metadata-cache>`
+
+.. _store-grib-fields-in-memory:
+
+store-grib-fields-in-memory
+++++++++++++++++++++++++++++
+
+When ``store-grib-fields-in-memory`` is ``True`` (this is the default) once a field is created it will be stored in the fieldlist. Otherwise the field will be created on demand and deleted when going out of scope.
+
+The actual memory used by a field depends on whether it stores the GRIB handle of the related GRIB message. This is controlled by the :ref:`grib-handle-cache-size <grib-handle-cache-size>` settings:
+
+ - When ``grib-handle-cache-size > 0`` the field objects themselves are lightweight and only store the GRIB handle cache index, while the actual GRIB handles are stored in the cache, which is attached to the fieldlist.
+ - When ``grib-handle-cache-size == 0`` the behaviour depends on ``store-grib-fields-in-memory``:
+
+    - when ``store-grib-fields-in-memory`` is ``True`` the fields does not own their GRIB handle but for each call to data and metadata access a new GRIB handle is created and released once the access finished. This can be useful when the fields have to be kept in memory but the memory usage should be low.
```

Comment:
Is this the right way round? This sounds like what I'd expect if the flag was False

Reply:
Following up our discussion the setting were completely refactored.